File: _engine/java/v0_1_0/Mock.java
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

package aifunc._engine.java.v0_1_0;

import aifunc.AIFuncConfig;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Mock data lookup and schema-based value generation. */
public final class Mock {

    private Mock() {}

    /**
     * Finds a matching mock output for the given input.
     *
     * <p>Lookup order:
     * <ol>
     *   <li>First entry whose {@code input} field deep-equals the actual input
     *   <li>First entry with no {@code input} field (fallback default)
     *   <li>{@code null} when no entries match
     * </ol>
     */
    public static Map<String, Object> findMockOutput(List<Types.MockEntry> entries, Map<String, Object> input) {
        Map<String, Object> fallback = null;
        for (Types.MockEntry entry : entries) {
            if (entry.getInput() == null) {
                if (fallback == null) fallback = entry.getOutput();
                continue;
            }
            if (Validator.deepEqual(entry.getInput(), input)) {
                return entry.getOutput();
            }
        }
        return fallback;
    }

    /**
     * Generates a zero-value instance conforming to the given JSON Schema.
     *
     * <ul>
     *   <li>Returns the {@code default} value if present
     *   <li>Returns the first {@code enum} value if present
     *   <li>Otherwise generates: {@code ""} / {@code 0.0} / {@code false} /
     *       {@code []} / {@code {}} based on type
     * </ul>
     */
    @SuppressWarnings("unchecked")
    public static Object generateFromSchema(Map<String, Object> schema) {
        if (schema == null) return null;

        if (schema.containsKey("default")) return schema.get("default");

        Object enumVal = schema.get("enum");
        if (enumVal instanceof List) {
            List<?> arr = (List<?>) enumVal;
            if (!arr.isEmpty()) return arr.get(0);
        }

        String type = resolveType(schema);
        switch (type) {
            case "string": {
                Object desc = schema.get("description");
                return (desc instanceof String) ? (String) desc : "";
            }
            case "number":
            case "integer":
                return 0.0;
            case "boolean":
                return Boolean.FALSE;
            case "null":
                return null;
            case "array": {
                Object items = schema.get("items");
                if (items instanceof Map) {
                    List<Object> list = new ArrayList<>();
                    list.add(generateFromSchema((Map<String, Object>) items));
                    return list;
                }
                return new ArrayList<>();
            }
            case "object": {
                Map<String, Object> obj = new LinkedHashMap<>();
                Object propsVal = schema.get("properties");
                if (propsVal instanceof Map) {
                    for (Map.Entry<?, ?> e : ((Map<?, ?>) propsVal).entrySet()) {
                        if (e.getValue() instanceof Map) {
                            obj.put(String.valueOf(e.getKey()),
                                    generateFromSchema((Map<String, Object>) e.getValue()));
                        }
                    }
                }
                return obj;
            }
            default:
                return null;
        }
    }

    /**
     * Resolves mock entries from an {@link AIFuncConfig}'s {@code mockData} field.
     * Accepts a {@code List<Types.MockEntry>}, a raw {@code List<Map>}, or a
     * {@code Map} with a {@code "cases"} key.
     */
    @SuppressWarnings("unchecked")
    public static List<Types.MockEntry> resolveMockEntries(Object mockData) {
        List<Types.MockEntry> result = new ArrayList<>();
        if (mockData == null) return result;

        if (mockData instanceof List) {
            for (Object item : (List<?>) mockData) {
                if (item instanceof Types.MockEntry) {
                    result.add((Types.MockEntry) item);
                } else if (item instanceof Map) {
                    result.add(Types.MockEntry.fromMap((Map<String, Object>) item));
                }
            }
            return result;
        }

        if (mockData instanceof Map) {
            Object casesVal = ((Map<?, ?>) mockData).get("cases");
            if (casesVal instanceof List) {
                for (Object item : (List<?>) casesVal) {
                    if (item instanceof Types.MockEntry) {
                        result.add((Types.MockEntry) item);
                    } else if (item instanceof Map) {
                        result.add(Types.MockEntry.fromMap((Map<String, Object>) item));
                    }
                }
            }
        }
        return result;
    }

    private static String resolveType(Map<String, Object> schema) {
        Object typeVal = schema.get("type");
        if (typeVal instanceof String) return (String) typeVal;
        if (typeVal instanceof List) {
            List<?> list = (List<?>) typeVal;
            if (!list.isEmpty() && list.get(0) instanceof String) return (String) list.get(0);
        }
        return "";
    }
}

File: _engine/csharp/v0_1_0/Mock.cs
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections;
using System.Collections.Generic;

namespace Aifunc.Engine.Csharp.V0_1_0;

/// <summary>Mock data lookup and schema-based value generation.</summary>
public static class Mock
{
    /// <summary>
    /// Finds a matching mock output for the given input.
    ///
    /// Lookup order:
    /// <list type="number">
    ///   <item>First entry whose <c>input</c> field deep-equals the actual input</item>
    ///   <item>First entry with no <c>input</c> field (fallback default)</item>
    ///   <item><c>null</c> when no entries match</item>
    /// </list>
    /// </summary>
    public static Dictionary<string, object?>? FindMockOutput(
        List<Types.MockEntry> entries, Dictionary<string, object?> input)
    {
        Dictionary<string, object?>? fallback = null;
        foreach (var entry in entries)
        {
            if (entry.Input is null)
            {
                if (fallback is null) fallback = entry.Output;
                continue;
            }
            if (Validator.DeepEqual(entry.Input, input))
            {
                return entry.Output;
            }
        }
        return fallback;
    }

    /// <summary>
    /// Generates a zero-value instance conforming to the given JSON Schema.
    ///
    /// <list type="bullet">
    ///   <item>Returns the <c>default</c> value if present</item>
    ///   <item>Returns the first <c>enum</c> value if present</item>
    ///   <item>Otherwise generates: <c>""</c> / <c>0.0</c> / <c>false</c> /
    ///         <c>[]</c> / <c>{}</c> based on type</item>
    /// </list>
    /// </summary>
    public static object? GenerateFromSchema(Dictionary<string, object?>? schema)
    {
        if (schema is null) return null;

        if (schema.ContainsKey("default")) return schema["default"];

        if (schema.TryGetValue("enum", out var enumVal) && enumVal is System.Collections.IList arr
            && arr.Count > 0)
        {
            return arr[0];
        }

        var type = ResolveType(schema);
        switch (type)
        {
            case "string":
            {
                if (schema.TryGetValue("description", out var desc) && desc is string ds)
                    return ds;
                return "";
            }
            case "number":
            case "integer":
                return 0.0;
            case "boolean":
                return false;
            case "null":
                return null;
            case "array":
            {
                if (schema.TryGetValue("items", out var items) && AsDict(items) is { } itemSchema)
                {
                    return new List<object?> { GenerateFromSchema(itemSchema) };
                }
                return new List<object?>();
            }
            case "object":
            {
                var obj = new Dictionary<string, object?>();
                if (schema.TryGetValue("properties", out var propsVal) && AsDict(propsVal) is { } props)
                {
                    foreach (var e in props)
                    {
                        if (AsDict(e.Value) is { } propSchema)
                        {
                            obj[e.Key] = GenerateFromSchema(propSchema);
                        }
                    }
                }
                return obj;
            }
            default:
                return null;
        }
    }

    /// <summary>
    /// Resolves mock entries from an <c>AIFuncConfig</c>'s <c>MockData</c> field.
    /// Accepts a <c>List&lt;Types.MockEntry&gt;</c>, a raw <c>List</c> of maps, or a
    /// map with a <c>"cases"</c> key.
    /// </summary>
    public static List<Types.MockEntry> ResolveMockEntries(object? mockData)
    {
        var result = new List<Types.MockEntry>();
        if (mockData is null) return result;

        if (mockData is System.Collections.IList list)
        {
            foreach (var item in list)
            {
                AddEntry(result, item);
            }
            return result;
        }

        if (mockData is System.Collections.IDictionary map)
        {
            if (map.Contains("cases"))
            {
                var casesVal = map["cases"];
                if (casesVal is System.Collections.IList cases)
                {
                    foreach (var item in cases)
                    {
                        AddEntry(result, item);
                    }
                }
            }
        }
        return result;
    }

    private static void AddEntry(List<Types.MockEntry> result, object? item)
    {
        if (item is Types.MockEntry entry)
        {
            result.Add(entry);
        }
        else if (AsDict(item) is { } map)
        {
            result.Add(Types.MockEntry.FromMap(map));
        }
    }

    private static string ResolveType(Dictionary<string, object?> schema)
    {
        if (!schema.TryGetValue("type", out var typeVal) || typeVal is null) return "";
        if (typeVal is string s) return s;
        if (typeVal is System.Collections.IList list && list.Count > 0 && list[0] is string first)
            return first;
        return "";
    }

    private static Dictionary<string, object?>? AsDict(object? v)
    {
        if (v is Dictionary<string, object?> d) return d;
        if (v is System.Collections.IDictionary raw)
        {
            var converted = new Dictionary<string, object?>();
            foreach (System.Collections.DictionaryEntry e in raw)
            {
                converted[Convert.ToString(e.Key) ?? ""] = e.Value;
            }
            return converted;
        }
        return null;
    }
}
